golf2/src/aes.rs

```diff
@@ -0,0 +1,269 @@
+use core::cell::Cell;
+use hotel::crypto::aes::AesEngine;
+use hotel::hil::aes::{AesClient, Interrupt, KeySize};
+use hotel::hil::common::SyscallError;
+use kernel::{AppId, Callback, Driver, Container, Shared, AppSlice};
+
+#[derive(Default)]
+struct Callbacks {
+    wfifo_overflow: Option<Callback>,
+    rfifo_overflow: Option<Callback>,
+    rfifo_underflow: Option<Callback>,
+    done_cipher: Option<Callback>,
+    done_key_expansion: Option<Callback>,
+    done_wipe_secrets: Option<Callback>,
+}
+
+#[derive(Default)]
+pub struct AppData {
+    key: Option<AppSlice<Shared, u8>>,
+    input_buffer: Option<AppSlice<Shared, u8>>,
+    output_buffer: Option<AppSlice<Shared, u8>>,
+    callbacks: Callbacks,
+}
+
+pub struct AesDriver<'a> {
+    device: &'a AesEngine,
+    apps: Container<AppData>,
+    current_user: Cell<Option<AppId>>,
+}
+
+impl<'a> AesDriver<'a> {
+    pub fn new(device: &'a mut AesEngine, container: Container<AppData>) -> AesDriver<'a> {
+        AesDriver {
+            device: device,
+            apps: container,
+            current_user: Cell::new(None),
+        }
+    }
+
+    fn setup(&self, caller_id: AppId, key_size: usize) -> Result<isize, SyscallError> {
+        self.apps
+            .enter(caller_id, |app_data, _| {
+                if self.current_user.get().is_some() {
+                    return Err(SyscallError::ResourceBusy);
+                }
+                self.current_user.set(Some(caller_id));
```

Comment:
This seems like it's going to leave the driver in a bad state if the argument checks fail later on, since it'll be held by a user even if the actual device hasn't been setup. The (side-effect-free) argument checks should be made before everything else as much as possible.